lib/record_store/provider/dnsimple/patch_api_header.rb

```diff
@@ -0,0 +1,26 @@
+# Patch Dnsimple client method which retrieves headers for API rate limit dynamically
+module Dnsimple
+  class Client
+    def execute(method, path, data = nil, options = {})
+      response = request(method, path, data, options)
+      rate_limit_sleep(response.headers["x-ratelimit-reset"].to_i, response.headers["x-ratelimit-remaining"].to_i)
+
+      case response.code
+      when 200..299
+        response
+      when 401
+        raise AuthenticationFailed, response["message"]
+      when 404
+        raise NotFoundError, response
+      else
+        raise RequestError, response
+      end
+    end
+
+    private
+
+    def rate_limit_sleep(rate_limit_reset, rate_limit_remaining)
+      sleep([0, rate_limit_reset - Time.now.to_i].max / (rate_limit_remaining + 1).to_f)
```

Comment:
Wondering if we could avoid to call `sleep` when the result is `0`
Something like:

```suggestion
wait_time = [0, rate_limit_reset - Time.now.to_i].max / (rate_limit_remaining + 1).to_f

sleep(wait_time) if wait_time > 0
```

Would of course require you to adapt the tests below to make sure that `sleep` is **not called* when the `wait_time` is computed to be `0`

Reply:
Gotcha, that makes sense to do. I'll add in this change.

Reply:
Another helpful thing (at least for the tests) would be to return a boolean depending on what that method does:

```
return false unless wait_time > 0
sleep(wait_time)
return true
```

Which would allow to directly check that the method return value is `false` when you expect not to have to wait

Reply:
Ah ya after looking a bit more into this, it seems that `MiniTest` doesn't exactly have something to test if a function "is not called". Some ideas were to assert an error and check if error is raised, similar to checking for bool as you mentioned, but that would mean having to catch the error whenever `rate_limit_sleep` is called. I'll try out returning a bool.